internal/backend/runtime/omni/controllers/omni/discovery_service_config_patch.go

```diff
@@ -0,0 +1,96 @@
+// included in the LICENSE file.
+
+package omni
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic/qtransform"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v3"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/internal/pkg/siderolink"
+)
+
+// DiscoveryServiceConfigPatchPrefix is the prefix for the system config patch that contains the discovery service endpoint config.
+const DiscoveryServiceConfigPatchPrefix = "950-discovery-service-"
+
+// DiscoveryServiceConfigPatchController creates the system config patch that contains the discovery service endpoint config.
+//
+// It sets the cluster discovery service endpoint to the embedded discovery service if the feature is available for the Omni instance and enabled for the cluster.
+type DiscoveryServiceConfigPatchController = qtransform.QController[*omni.Cluster, *omni.ConfigPatch]
+
+// NewDiscoveryServiceConfigPatchController initializes DiscoveryServiceConfigPatchController.
+func NewDiscoveryServiceConfigPatchController(enabled bool, port int) *DiscoveryServiceConfigPatchController {
+	getPatch := sync.OnceValues(func() (string, error) {
+		var sb strings.Builder
+
+		encoder := yaml.NewEncoder(&sb)
+
+		encoder.SetIndent(2)
+
+		if err := encoder.Encode(map[string]any{
+			"cluster": map[string]any{
+				"discovery": map[string]any{
+					"registries": map[string]any{
+						"service": map[string]any{
+							"endpoint": "http://" + net.JoinHostPort(siderolink.ListenHost, strconv.Itoa(port)),
+						},
+					},
+				},
+			},
+		}); err != nil {
+			return "", fmt.Errorf("failed to encode patch: %w", err)
+		}
+
+		return sb.String(), nil
+	})
+
+	return qtransform.NewQController(
+		qtransform.Settings[*omni.Cluster, *omni.ConfigPatch]{
+			Name: "DiscoveryServiceConfigPatchController",
+			MapMetadataFunc: func(cluster *omni.Cluster) *omni.ConfigPatch {
+				return omni.NewConfigPatch(resources.DefaultNamespace, DiscoveryServiceConfigPatchPrefix+cluster.Metadata().ID())
+			},
+			UnmapMetadataFunc: func(configPatch *omni.ConfigPatch) *omni.Cluster {
+				id := strings.TrimPrefix(configPatch.Metadata().ID(), DiscoveryServiceConfigPatchPrefix)
+
+				return omni.NewCluster(resources.DefaultNamespace, id)
+			},
+			TransformFunc: func(_ context.Context, _ controller.Reader, _ *zap.Logger, cluster *omni.Cluster, configPatch *omni.ConfigPatch) error {
+				configPatch.Metadata().Labels().Set(omni.LabelSystemPatch, "")
+				configPatch.Metadata().Labels().Set(omni.LabelCluster, cluster.Metadata().ID())
+
+				enabledForCluster := cluster.TypedSpec().Value.GetFeatures().GetUseEmbeddedDiscoveryService()
+
+				if !enabled || !enabledForCluster {
+					configPatch.TypedSpec().Value.Data = ""
+
+					return nil
```

Comment:
I think we should use the destroy tag here to drop the patch or skip the creation of the patch.

Reply:
Yes, that was the reason I added it, but we decided to hold it for now: https://github.com/cosi-project/runtime/pull/452#discussion_r1622292823

Reply:
Well, I don't think it's just a quality of life improvement :)
So I would suggest no to hold-off and add destroy tag support.

Reply:
there's another place like that, which might benefit from that as well.
For now I did in a bit hacky way:
https://github.com/siderolabs/omni/blob/c6358272732628b01e38efdccbcfd766205ed7da/internal/backend/runtime/omni/controllers/omni/maintenance_config_patch.go#L57

Reply:
Not sure if teardown & check if ready & destroy is the right way either atm.

Reply:
Only weird thing here is that we don't remove the finalizer from the main input even though we didn't create the resource, otherwise it works

Reply:
will use https://github.com/cosi-project/runtime/pull/452 to address this, when I get that right and get it merged.

Reply:
replaced the logic both here and in the `maintenance_config_patch.go` with the destroy output feature merged in cosi: https://github.com/cosi-project/runtime/commit/15e9d678159d409649a1afe95f4d225e0f0bcef8